### ISSUE TITLE
BGhostview hangs with no displayed content.

### ISSUE THREAD

**Issue by OscarL** [haikuports/haikuports#8518]
This is with BGhostview 20190401-6, on beta4 32 bits.

Tried to open a .ps file, it shows the correct list of pages on the right side of the window, but no content on the right side. Attempting to change page, or even close BGhostview's window, causes it to hang/freeze with little to no CPU usage (saw a `BLooper::ReadRawFromPort()` on Debugger, but not sure how relevant that is).

No info is printed to stdout when invoking it from Terminal either.

**Comment by pulkomandy** [comment]
Is this a regression from the migration to ghostscript 10?

**Comment by jmairboeck** [comment]
I tested the old version and saw the same behaviour.

**Comment by pulkomandy** [comment]
I think this should better be reported at github.com/haikuarchives/bghostview

Anyways, here are some investigations:

- I had to make some fixes for 64bit compatibility first, I will push these
- The code is architectured using semaphores to synchronize between the ghostscript thread and the UI thread. In particular, there is a painter_sem that's supposed to be acquired when Ghostscript sets the page size, and released when the page is ready to display. However, the gsapi_display_presize function is called twice and so it acquires the semaphore twice. The second time, it blocks because no one has released the semaphore
- The memory allocation in gsapi_memalloc needs a bit more memory than what we allocate for the bitmap (notice how we ignore the passed size parameter and do our own thing...) I think it store pointers to each line at the end of the structure. So the bitmap needs a bit of extra space for that (or, really, should we be using a BBitmap this way here?). I changed height - 1 to height + 2 in the BBitmap constructor parameters and in the case I was testing with, it avoided this crash.
- I have temporarily commented out the semaphore (happily running without any sync), then the first page displays. But then it crashes when trying to open any other page, because re-initializing Ghostscript fails.

Overall it looks like my code here is just terribly broken

**Comment by pulkomandy** [comment]
I have made several fixes for the problems above:

- Fixed 64 bit issues
- Rewrote locking and allocation of the BBitmap, using a BLocker and an area

However it still crashes when trying to change pages. I tried to debug a bit with libroot_debug and the guarded heap, and it seems the issue is an use-after-free, with the allocation happening in the init of ghostscript gsapi_init_with_args calling gmtime_r which calls tzset which results in some allocation in ICU, and the use after free in gs_lib_ctx_init. At this point I don't think the bug is in BGhostView, it seems to be either in ghostscript or in ICU? Unless I missed something of course.

I hope it still works better for you and at least this version can be enabled for x86_64 since it will compile there and run a little before crashing.
